datasource/opennms/src/test/java/org/opennms/oce/datasource/opennms/processors/InventoryTableProcessorTest.java

```diff
@@ -0,0 +1,172 @@
+
+    @Test
+    public void testAddRemoveCallbacks() {
+        long step = 10000L;
+
+        KeyValueStore inventoryStore = mock(KeyValueStore.class);
+        //noinspection unchecked
+        when(inventoryStore.delete(any(String.class))).thenAnswer((Answer) invocation ->
+                inventoryMap.remove((String) invocation.getArguments()[0]));
+
+        ProcessorContext mockProcessorContext = mock(ProcessorContext.class);
+        when(mockProcessorContext.getStateStore(any(String.class))).thenReturn(inventoryStore);
+
+        HandlerRegistry<InventoryHandler> handlerRegistry = new HandlerRegistry<>();
+        handlerRegistry.register(new TestInventoryHandler());
+
+        InventoryTableProcessor inventoryTableProcessor = new InventoryTableProcessor(handlerRegistry, step, step);
+        inventoryTableProcessor.init(mockProcessorContext);
+
+        ArgumentCaptor<Punctuator> punctuatorArgumentCaptor = ArgumentCaptor.forClass(Punctuator.class);
+        verify(mockProcessorContext).schedule(any(Long.class), any(PunctuationType.class),
+                punctuatorArgumentCaptor.capture());
+        Punctuator capturedPunctuator = punctuatorArgumentCaptor.getValue();
+
+        // Create two identical inventory with different owners and different expiry times
+        inventoryMap.put("owner1", getInventoryObjectExpiringAt(step));
+        inventoryMap.put("owner2", getInventoryObjectExpiringAt(step * 2));
+        inventoryMap.forEach(inventoryTableProcessor::process);
+        // Two of the same inventory should result in only one add
+        verifyCalledNTimes(addCallbackCounter, 1);
+
+        // Since the inventory object had two references the first delete should not result in handlers being notified
+        // of a remove
+        when(inventoryStore.all()).thenReturn(new MockKeyValueIteratorImpl(inventoryMap));
+        capturedPunctuator.punctuate(step);
+        verifyCalledNTimes(removeCallbackCounter, 0);
+
+        // The second delete should result in the inventory object no longer being referenced and now the handler should
+        // be notified
+        when(inventoryStore.all()).thenReturn(new MockKeyValueIteratorImpl(inventoryMap));
+        capturedPunctuator.punctuate(step * 2);
+        verifyCalledNTimes(removeCallbackCounter, 1);
+    }
+
+    private InventoryModelProtos.InventoryObjects getInventoryObjectExpiringAt(long expiryTime) {
+        return InventoryModelProtos.InventoryObjects
+                .newBuilder()
+                .addInventoryObject(InventoryModelProtos.InventoryObject.newBuilder()
+                        .setId("test.id")
+                        .build())
+                .setExpiresAt(expiryTime)
+                .build();
+    }
+
+    private void verifyCalledNTimes(AtomicInteger callbackCounter, int numTimes) {
+        await().atMost(1, TimeUnit.SECONDS).until(() -> callbackCounter.get() == numTimes);
```

Comment:
Do we need to use `await()` here? Nothing appears to be asynchronous in this test.

Reply:
Looks like we don't need to. Brain gets stuck in async mode when looking at callbacks...